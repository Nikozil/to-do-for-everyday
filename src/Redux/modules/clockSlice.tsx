import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { getTime } from 'date-fns';
import { AppThunk } from '../store';

export const initialState = {
  time: getTime(new Date()) as number,
  interval: undefined as number | undefined,
};

export type initialStateType = typeof initialState;
const clockSlice = createSlice({
  name: 'clock',
  initialState,
  reducers: {
    TICK: (state) => {
      return {
        ...state,
        time: getTime(new Date()),
      };
    },
    setClockInterval: (state, action: PayloadAction<any>) => {
      return {
        ...state,
        interval: action.payload,
      };
    },
  },
});

export const { TICK, setClockInterval } = clockSlice.actions;

export const startClock = (): AppThunk => (dispatch, getState) => {
  dispatch(stopClock());

  const interval = setInterval(() => {
    dispatch(TICK());
  }, 1000);

  dispatch(setClockInterval(interval));
};
export const stopClock = (): AppThunk => (dispatch, getState) => {
  clearInterval(getState().clock.interval);
  dispatch(setClockInterval(null));
};

export default clockSlice;
