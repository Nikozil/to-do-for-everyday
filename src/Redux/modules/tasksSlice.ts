import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import {
  add,
  endOfToday,
  format,
  getTime,
  milliseconds,
  startOfDay,
} from 'date-fns';

import { StoreAPI } from '../../api/StoreAPI';
import { AppThunk } from '../store';

export const initialState = {
  tasksList: [] as Task[],
  livedDay: {
    score: 0,
    tag: '',
    doneTasksList: [],
    timestamp: null,
  } as LivedDay,
  initStatus: false as boolean,
};

export type initialStateType = typeof initialState;
const tasksSlice = createSlice({
  name: 'tasks',
  initialState,
  reducers: {
    setTasks: (state, action: PayloadAction<Task[]>) => {
      return { ...state, tasksList: action.payload };
    },
    appendNewTask: (state, action: PayloadAction<Task>) => {
      state.tasksList.push(action.payload);
    },
    editTask: (state, action: PayloadAction<PartialTask>) => {
      const task = state.tasksList.find(
        (task) => task.id === action.payload.id
      );
      if (task) {
        task.data = { ...task.data, ...action.payload.data };
      }
    },
    removeTask: (state, action: PayloadAction<string>) => {
      state.tasksList = state.tasksList.filter(
        (task) => task.id !== action.payload
      );
    },
    setInitStatus: (state, action: PayloadAction<boolean>) => {
      state.initStatus = action.payload;
    },
    setLivedDay: (state, action: PayloadAction<LivedDay>) => {
      state.livedDay = { ...state.livedDay, ...action.payload };
    },
    appendTaskToDoneTasksList: (state, action: PayloadAction<LivedTask>) => {
      state.livedDay.doneTasksList.push(action.payload);
    },
    removeTaskToDoneTasksList: (state, action: PayloadAction<string>) => {
      const newdoneTasksList = state.livedDay.doneTasksList.filter(
        (task) => task.id !== action.payload
      );
      state.livedDay.doneTasksList = newdoneTasksList;
    },
    setTag: (state, action: PayloadAction<string>) => {
      state.livedDay.tag = action.payload;
    },
    setScore: (state, action: PayloadAction<Score>) => {
      state.livedDay.score = action.payload;
    },
    setTimestamp: (state, action: PayloadAction<number>) => {
      state.livedDay.timestamp = action.payload;
    },
  },
});

export const {
  setTasks,
  appendNewTask,
  editTask,
  removeTask,
  setInitStatus,
  setLivedDay,
  appendTaskToDoneTasksList,
  removeTaskToDoneTasksList,
  setTag,
  setScore,
  setTimestamp,
} = tasksSlice.actions;

export const getTasks = (): AppThunk => async (dispatch, getState) => {
  try {
    let tasks = await StoreAPI.getTask();
    let time = getState().clock.time;

    let livedDay = await StoreAPI.getLivedDay(format(time, 'dd.MM.yyyy'));

    if (!tasks) throw new Error('Не удалось получить данные');
    if (!livedDay) throw new Error('Не удалось получить данные');

    dispatch(setTasks(tasks));
    dispatch(setLivedDay(livedDay));
    dispatch(setInitStatus(true));
  } catch (err: any) {
    return err.message as string;
  }
};
export const addTask =
  (taskName: string, duration: Duration, repeat = 0): AppThunk =>
  async (dispatch, getState) => {
    const time = getTime(add(getState().clock.time, duration));
    const taskData = { name: taskName, done: false, time, repeat } as TaskData;
    try {
      const newTask = await StoreAPI.setTask(taskData);
      if (!newTask) throw new Error('Не удалось создать задачу');
      dispatch(appendNewTask(newTask));
    } catch (err: any) {
      return err.message as string;
    }
  };
export const updateTask =
  (id: string, data: PartialTaskData): AppThunk =>
  async (dispatch) => {
    try {
      await StoreAPI.updateTask(id, data);
      dispatch(editTask({ id, data }));
    } catch (err: any) {
      return err.message as string;
    }
  };
export const checkTask =
  (task: Task): AppThunk =>
  async (dispatch, getState) => {
    try {
      const { id, data } = task;
      const { name, done, time, repeat } = data;
      let newData;
      // make the data repeat date or complete
      if (repeat) {
        //move task for future
        let newTime = time;
        //count expiration of task + 1 repeat
        let expiration = Math.round(
          (getTime(endOfToday()) - newTime) /
            (milliseconds({ days: 1 }) * repeat) +
            1
        );
        //count date of future task
        newTime = getTime(add(newTime, { days: repeat * expiration }));
        newData = { time: newTime };
      } else {
        newData = { done: !done };
      }
      // add task to done list
      dispatch(checkupTimestamp());
      const doneTasks = {
        id: id,
        name: name,
      };
      //send data as batch
      const taskApiData = { taskId: id, taskData: newData };
      const doneTaskApiData = {
        doneTaskDate: format(getState().clock.time, 'dd.MM.yyyy'),
        doneTask: doneTasks,
        doneTaskRemove: false,
      };
      await StoreAPI.setBatchDoneTask(taskApiData, doneTaskApiData);
      //update store
      dispatch(editTask({ id, data: newData }));
      dispatch(appendTaskToDoneTasksList(doneTasks));
    } catch (err: any) {
      return err.message as string;
    }
  };
export const uncheckTask =
  (task: LivedTask): AppThunk =>
  async (dispatch, getState) => {
    const { id, name } = task;
    try {
      let task = getState().tasks.tasksList.find((task) => task.id === id);
      if (task) {
        const { done, time, repeat } = task.data;
        let newData;
        // make the data repeat date or complete
        if (repeat) {
          //move date to past
          newData = { time: getTime(add(time, { days: -repeat })) };
        } else {
          //return task time and status
          newData = done
            ? { done: !done }
            : { done: false, time: getState().clock.time };
        }
        const taskApiData = { taskId: id, taskData: newData };
        const doneTask = {
          id: id,
          name: name,
        };
        const doneTaskApiData = {
          doneTaskDate: format(getState().clock.time, 'dd.MM.yyyy'),
          doneTask: doneTask,
          doneTaskRemove: true,
        };
        await StoreAPI.setBatchDoneTask(taskApiData, doneTaskApiData);

        //update store

        dispatch(editTask({ id, data: newData }));
        dispatch(removeTaskToDoneTasksList(id));
      } else throw new Error('Нет такой задачи');
    } catch (err: any) {
      return err.message as string;
    }
  };
export const deleteTask =
  (id: string): AppThunk =>
  async (dispatch) => {
    try {
      await StoreAPI.deleteTask(id);
      dispatch(removeTask(id));
    } catch (err: any) {
      return err.message as string;
    }
  };
export const addTag =
  (tag: string): AppThunk =>
  async (dispatch, getState) => {
    try {
      dispatch(checkupTimestamp());
      const time = format(getState().clock.time, 'dd.MM.yyyy');
      const livedDay = { tag };
      await StoreAPI.updateLivedDay(time, livedDay);
      dispatch(setTag(tag));
    } catch (err: any) {
      return err.message as string;
    }
  };
export const addScore =
  (score: Score): AppThunk =>
  async (dispatch, getState) => {
    try {
      dispatch(checkupTimestamp());
      const date = format(getState().clock.time, 'dd.MM.yyyy');
      const livedDay = { score };
      await StoreAPI.updateLivedDay(date, livedDay);
      dispatch(setScore(score));
    } catch (err: any) {
      return err.message as string;
    }
  };
export const checkupTimestamp = (): AppThunk => async (dispatch, getState) => {
  if (!getState().tasks.livedDay.timestamp) {
    try {
      const date = format(getState().clock.time, 'dd.MM.yyyy');

      const timestamp = getTime(startOfDay(getState().clock.time));
      const livedDay = { timestamp };
      await StoreAPI.updateLivedDay(date, livedDay);
      dispatch(setTimestamp(timestamp));
    } catch (err: any) {
      return err.message as string;
    }
  }
};

export const doAgainTask =
  (task: Task): AppThunk =>
  async (dispatch, getState) => {
    try {
      const { id, data } = task;
      const { done } = data;

      //set date of tomorrow task
      const currentTime = getState().clock.time;
      const newTime = getTime(add(currentTime, { days: 1 }));

      const newData = { time: newTime, done: !done };

      await StoreAPI.updateTask(id, newData);
      //update store
      dispatch(editTask({ id, data: newData }));
    } catch (err: any) {
      return err.message as string;
    }
  };

export default tasksSlice;

export interface Task {
  id: string;
  data: TaskData;
}
export interface TaskData {
  name: string;
  done: boolean;
  time: number;
  repeat: number;
}
export type PartialTaskData = Partial<TaskData>;
export interface PartialTask {
  id: string;
  data: PartialTaskData;
}
export interface LivedTask {
  id: string;
  name: string;
}
export interface LivedDay {
  tag: string;
  score: Score;
  doneTasksList: LivedTask[];
  timestamp: null | number;
}
export type Score = 0 | 1 | 2 | 3 | 4 | 5;
