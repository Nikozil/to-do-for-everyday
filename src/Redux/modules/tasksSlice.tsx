import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { add, endOfToday, format, getTime, milliseconds } from 'date-fns';
import { StoreAPI } from '../../api/StoreAPI';
import { AppThunk } from '../store';

export const initialState = {
  tasksList: [] as Task[],
  doneTasksList: {} as DoneTask,
  initStatus: false as boolean,
};

export type initialStateType = typeof initialState;
const tasksSlice = createSlice({
  name: 'tasks',
  initialState,
  reducers: {
    setTasks: (state, action: PayloadAction<Task[]>) => {
      return { ...state, tasksList: action.payload };
    },
    editTask: (state, action: PayloadAction<PartialTask>) => {
      const task = state.tasksList.find(
        (task) => task.id === action.payload.id
      );
      if (task) {
        task.data = { ...task.data, ...action.payload.data };
      }
    },
    removeTask: (state, action: PayloadAction<string>) => {
      state.tasksList = state.tasksList.filter(
        (task) => task.id !== action.payload
      );
    },
    setInitStatus: (state, action: PayloadAction<boolean>) => {
      state.initStatus = action.payload;
    },
    setDoneTasks: (state, action: PayloadAction<DoneTask>) => {
      return { ...state, doneTasksList: action.payload };
    },
    addTaskToDoneTasksList: (state, action: PayloadAction<DoneTask>) => {
      return {
        ...state,
        doneTasksList: { ...state.doneTasksList, ...action.payload },
      };
    },
    removeTaskToDoneTasksList: (state, action: PayloadAction<string>) => {
      delete state.doneTasksList[action.payload];
    },
  },
});

export const {
  setTasks,
  editTask,
  removeTask,
  setInitStatus,
  setDoneTasks,
  addTaskToDoneTasksList,
  removeTaskToDoneTasksList,
} = tasksSlice.actions;

export const getTasks = (): AppThunk => async (dispatch) => {
  try {
    let tasks = await StoreAPI.getTask();
    let doneTasks = await StoreAPI.getDoneTask(
      format(new Date(), 'dd.MM.yyyy')
    );

    if (!tasks) throw new Error('Не удалось получить данные');
    if (!doneTasks) throw new Error('Не удалось получить данные');

    dispatch(setTasks(tasks));
    dispatch(setDoneTasks(doneTasks));
    dispatch(setInitStatus(true));
  } catch (err: any) {
    return err.message as string;
  }
};
export const addTask =
  (taskName: string, time: number, repeat = 0): AppThunk =>
  async (dispatch) => {
    const taskData = { name: taskName, done: false, time, repeat } as TaskData;
    try {
      await StoreAPI.setTask(taskData);
      dispatch(getTasks());
    } catch (err: any) {
      return err.message as string;
    }
  };
export const updateTask =
  (id: string, data: PartialTaskData): AppThunk =>
  async (dispatch) => {
    try {
      await StoreAPI.updateTask(id, data);
      dispatch(editTask({ id, data }));
    } catch (err: any) {
      return err.message as string;
    }
  };
export const checkTask =
  (task: Task): AppThunk =>
  async (dispatch, getState) => {
    try {
      const { id, data } = task;
      const { name, done, time, repeat } = data;
      let newData;
      // make the data repeat date or complete
      if (repeat) {
        //move task for future
        let newTime = time;
        //count expiration of task + 1 repeat
        let expiration = Math.round(
          (getTime(endOfToday()) - newTime) /
            (milliseconds({ days: 1 }) * repeat) +
            1
        );
        //count date of future task
        newTime = getTime(add(newTime, { days: repeat * expiration }));
        newData = { time: newTime };
      } else {
        newData = { done: !done };
      }
      // add task to done list
      const doneTasksList = {
        [id]: name,
      };
      //send data as batch
      const taskApiData = { taskId: id, taskData: newData };
      const doneTaskApiData = {
        doneTaskDate: format(new Date(), 'dd.MM.yyyy'),
        doneTasks: doneTasksList,
        doneTasksMerge: true,
      };
      await StoreAPI.setBatchDoneTask(taskApiData, doneTaskApiData);
      //update store
      dispatch(editTask({ id, data: newData }));
      dispatch(addTaskToDoneTasksList(doneTasksList));
    } catch (err: any) {
      return err.message as string;
    }
  };
export const uncheckTask =
  (id: string): AppThunk =>
  async (dispatch, getState) => {
    try {
      let task = getState().tasks.tasksList.find((task) => task.id === id);
      if (task) {
        const { done, time, repeat } = task.data;
        let newData;
        // make the data repeat date or complete
        if (repeat) {
          //move date to past
          newData = { time: getTime(add(time, { days: -repeat })) };
        } else {
          newData = { done: !done };
        }
        // add task to done list
        let doneTasksList = { ...getState().tasks.doneTasksList };
        delete doneTasksList[id];
        //send data as batch
        const taskApiData = { taskId: id, taskData: newData };
        const doneTaskApiData = {
          doneTaskDate: format(new Date(), 'dd.MM.yyyy'),
          doneTasks: doneTasksList,
          doneTasksMerge: false,
        };
        await StoreAPI.setBatchDoneTask(taskApiData, doneTaskApiData);

        //update store

        dispatch(editTask({ id, data: newData }));
        dispatch(removeTaskToDoneTasksList(id));
      } else throw new Error('Нет такой задачи');
    } catch (err: any) {
      return err.message as string;
    }
  };
export const deleteTask =
  (id: string): AppThunk =>
  async (dispatch) => {
    try {
      await StoreAPI.deleteTask(id);
      dispatch(removeTask(id));
    } catch (err: any) {
      return err.message as string;
    }
  };

export default tasksSlice;

export interface Task {
  id: string;
  data: TaskData;
}
export interface TaskData {
  name: string;
  done: boolean;
  time: number;
  repeat: number;
}
export type PartialTaskData = Partial<TaskData>;
interface PartialTask {
  id: string;
  data: PartialTaskData;
}
export interface DoneTask {
  [id: string]: string;
}
