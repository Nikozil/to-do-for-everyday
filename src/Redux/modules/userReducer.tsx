import { User } from '@firebase/auth/dist/auth-exp-public';
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { AuthAPI } from '../../api/AuthAPI';
import { UserAPI } from '../../api/UserAPI';
import { AppThunk } from '../store';

let initialState = {
  userData: { uid: '', displayName: null, email: null } as UserData,
  authStatus: false as boolean,
  initStatus: false as boolean,
  loginError: null as LoginErrorType,
};

export type initialStateType = typeof initialState;
const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    setUserData: (state, action: PayloadAction<User | UserData>) => {
      const { uid, email, displayName } = action.payload;
      return { ...state, userData: { uid, email, displayName } };
    },
    setAuthStatus: (state, action: PayloadAction<boolean>) => {
      state.authStatus = action.payload;
    },
    setInitStatus: (state, action: PayloadAction<boolean>) => {
      state.initStatus = action.payload;
    },
    setError: (state, action: PayloadAction<LoginErrorType>) => {
      state.loginError = action.payload;
    },
  },
});

export const { setUserData, setAuthStatus, setInitStatus, setError } =
  userSlice.actions;

export const signIn =
  (email: string, password: string, remember: boolean): AppThunk =>
  async (dispatch) => {
    try {
      await AuthAPI.signIn(email, password, remember);
      dispatch(setError(null));
    } catch (err: any) {
      dispatch(setError(err.message as string));
    }
  };
export const signOut = (): AppThunk => async (dispatch) => {
  await AuthAPI.signOut();
  dispatch(setAuthStatus(false));
  dispatch(setUserData(initialState.userData));
};
export const updateUserData = (): AppThunk => async (dispatch) => {
  await AuthAPI.updateUserStatus(
    dispatch,
    setUserData,
    setAuthStatus,
    setInitStatus
  );
};
export const updateProfile =
  (displayName: string): AppThunk =>
  async (dispatch) => {
    await UserAPI.updateProfile(displayName);
    dispatch(updateUserData());
  };

export default userSlice;

export type LoginErrorType = string | null;

export type UserData = Pick<User, 'uid' | 'displayName' | 'email'>;
