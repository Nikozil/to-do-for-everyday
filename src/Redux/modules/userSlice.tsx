import { User } from '@firebase/auth/dist/auth-exp-public';
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { AuthAPI } from '../../api/AuthAPI';
import { UserAPI } from '../../api/UserAPI';
import { AppThunk } from '../store';

let initialState = {
  userData: { uid: '', displayName: null, email: null } as UserData,
  authStatus: false as boolean,
  initStatus: false as boolean,
};

export type initialStateType = typeof initialState;
const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    setUserData: (state, action: PayloadAction<User | UserData>) => {
      const { uid, email, displayName } = action.payload;
      return { ...state, userData: { uid, email, displayName } };
    },
    setAuthStatus: (state, action: PayloadAction<boolean>) => {
      state.authStatus = action.payload;
    },
    setInitStatus: (state, action: PayloadAction<boolean>) => {
      state.initStatus = action.payload;
    },
  },
});

export const { setUserData, setAuthStatus, setInitStatus } = userSlice.actions;

export const signIn =
  (email: string, password: string, remember: boolean): AppThunk =>
  async (dispatch) => {
    try {
      await AuthAPI.signIn(email, password, remember);
      return 'Авторизация выполнена';
    } catch (err: any) {
      return err.message as string;
    }
  };
export const signOut = (): AppThunk => async (dispatch) => {
  try {
    await AuthAPI.signOut();
    dispatch(setAuthStatus(false));
    dispatch(setUserData(initialState.userData));
  } catch (err: any) {
    return err.message as string;
  }
};
export const updateUserData = (): AppThunk => async (dispatch) => {
  await AuthAPI.updateUserStatus(
    dispatch,
    setUserData,
    setAuthStatus,
    setInitStatus
  );
};
export const updateProfile =
  (displayName: string): AppThunk<Promise<string>> =>
  async (dispatch) => {
    try {
      await UserAPI.updateProfile(displayName);
    } catch (err: any) {
      return err.message;
    }
    dispatch(updateUserData());
  };
export const updatePassword =
  (oldPassword: string, newPassword: string): AppThunk<Promise<string>> =>
  async (dispatch) => {
    try {
      let responce = await AuthAPI.updatePassword(oldPassword, newPassword);
      return responce;
    } catch (err: any) {
      return err.message as string;
    }
  };

export default userSlice;

export type UserData = Pick<User, 'uid' | 'displayName' | 'email'>;
