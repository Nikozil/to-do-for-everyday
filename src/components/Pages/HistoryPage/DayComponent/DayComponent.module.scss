@import '../../../../App.scss';
.container {
  margin: 5px;
  padding: 5px;

  border-radius: 5px;

  // background-color: $background-color;
  // background: linear-gradient($background-color, #fff);

  box-shadow: 3px 3px 10px 1px $secondary-color;

  overflow: hidden;

  // background-color: $secondary-color;
  // color: $header-color;

  display: flex;
  flex-direction: column;
  justify-content: space-between;
  text-align: center;
}

.date {
  font-weight: 500;
  background-color: $secondary-color;
  color: $header-color;
  margin: -10px -10px 0 -10px;
  padding-top: 10px;
  padding-bottom: 15px;
  height: 40px;
}

.task-list {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 5px;
}
.task-li {
  list-style-type: none;
}

.tag {
  padding: 5px;
  margin-top: 5px;
  font-size: 0.8em;
  font-style: italic;
}
