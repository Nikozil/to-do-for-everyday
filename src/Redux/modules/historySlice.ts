import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { endOfMonth, getTime, startOfMonth } from 'date-fns';

import { StoreAPI } from '../../api/StoreAPI';
import { AppThunk } from '../store';
import { LivedDay } from './tasksSlice';

export const initialState = {
  daysList: [] as HistoryDay[],
  initStatus: false as boolean,
};

export type initialStateType = typeof initialState;
const historySlice = createSlice({
  name: 'history',
  initialState,
  reducers: {
    setDays: (state, action: PayloadAction<HistoryDay[]>) => {
      return { ...state, daysList: action.payload };
    },
    setInitStatus: (state, action: PayloadAction<boolean>) => {
      state.initStatus = action.payload;
    },
  },
});

export const { setDays, setInitStatus } = historySlice.actions;

export const getDays =
  (month: number, year: number): AppThunk =>
  async (dispatch, getState) => {
    try {
      dispatch(setInitStatus(false));
      const date = getTime(new Date(year, month));
      const start = getTime(startOfMonth(date));
      const end = getTime(endOfMonth(date));

      const days = await StoreAPI.getDays(start, end);
      if (!days) throw new Error('Не удалось получить данные');

      dispatch(setDays(days));
      dispatch(setInitStatus(true));
    } catch (err: any) {
      return err.message as string;
    }
  };

export default historySlice;

export interface HistoryDay extends LivedDay {
  timestamp: number;
}
