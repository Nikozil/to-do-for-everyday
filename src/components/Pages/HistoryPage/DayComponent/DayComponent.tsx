import React from 'react';
import { format } from 'date-fns';

import RaitingStarComponent from '../../../../common_components/RaitingStarComponent/RaitingStarComponent';
import { HistoryDay } from '../../../../Redux/modules/historySlice';
import { LivedTask } from '../../../../Redux/modules/tasksSlice';
import { FaCouch } from 'react-icons/fa';
import styles from './DayComponent.module.scss';

const DayComponent: React.FC<PropsType> = ({ day }) => {
  const { doneTasksList, score, timestamp, tag } = day;

  const isDoneTaskList = doneTasksList && doneTasksList.length;

  const date = format(timestamp, 'dd.MM.yyyy');

  const mapDoneTaskComponent = (task: LivedTask) => (
    <li className={styles['task-li']} key={task.id}>
      <span>{task.name}</span>
    </li>
  );

  const mapStars = (star: string, index: number) => {
    let scoreValue = index + 1;
    return (
      <RaitingStarComponent
        key={scoreValue}
        active={score >= scoreValue}
        size={15}
      />
    );
  };

  return (
    <section className={styles.container}>
      <h3 className={styles.date}>{date}</h3>
      <ul className={styles['task-list']}>
        {isDoneTaskList ? (
          doneTasksList.map(mapDoneTaskComponent)
        ) : (
          <FaCouch size={35} />
        )}
      </ul>
      <div>
        <span className={styles.tag}>{tag ?? null}</span>
        <div className={styles.score}>{[...Array(5)].map(mapStars)}</div>
      </div>
    </section>
  );
};

export default DayComponent;

interface PropsType {
  day: HistoryDay;
}
