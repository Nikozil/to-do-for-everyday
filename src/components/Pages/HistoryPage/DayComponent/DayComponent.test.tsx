import React from 'react';
import { HistoryDay } from '../../../../Redux/modules/historySlice';
import { render } from '../../../../utils/tests/test-utils';
import DayComponent from './DayComponent';

const testDay = {
  timestamp: 1635263950643,
  score: 3,
  tag: 'hi',
  doneTasksList: [{ name: 'do', id: '123' }],
} as HistoryDay;
describe('DayComponent', () => {
  it('render correct', async () => {
    const { asFragment } = render(<DayComponent day={testDay} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
