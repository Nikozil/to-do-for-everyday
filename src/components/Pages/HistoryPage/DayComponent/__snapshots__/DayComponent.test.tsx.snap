// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DayComponent render correct 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="date"
    >
      26.10.2021
    </div>
    <ul
      class="task-list"
    >
      <li
        class="task-li"
      >
        <div
          class="content"
        >
          <span>
            do
          </span>
        </div>
      </li>
    </ul>
    <div
      class="tag"
    >
      hi
    </div>
    <div
      class="score"
    >
      <svg
        fill="currentColor"
        height="15"
        stroke="currentColor"
        stroke-width="0"
        version="1.1"
        viewBox="0 0 16 16"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 6.204l-5.528-0.803-2.472-5.009-2.472 5.009-5.528 0.803 4 3.899-0.944 5.505 4.944-2.599 4.944 2.599-0.944-5.505 4-3.899z"
        />
      </svg>
      <svg
        fill="currentColor"
        height="15"
        stroke="currentColor"
        stroke-width="0"
        version="1.1"
        viewBox="0 0 16 16"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 6.204l-5.528-0.803-2.472-5.009-2.472 5.009-5.528 0.803 4 3.899-0.944 5.505 4.944-2.599 4.944 2.599-0.944-5.505 4-3.899z"
        />
      </svg>
      <svg
        fill="currentColor"
        height="15"
        stroke="currentColor"
        stroke-width="0"
        version="1.1"
        viewBox="0 0 16 16"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 6.204l-5.528-0.803-2.472-5.009-2.472 5.009-5.528 0.803 4 3.899-0.944 5.505 4.944-2.599 4.944 2.599-0.944-5.505 4-3.899z"
        />
      </svg>
      <svg
        fill="currentColor"
        height="15"
        stroke="currentColor"
        stroke-width="0"
        version="1.1"
        viewBox="0 0 16 16"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 6.204l-5.528-0.803-2.472-5.009-2.472 5.009-5.528 0.803 4 3.899-0.944 5.505 4.944-2.599 4.944 2.599-0.944-5.505 4-3.899zM8 11.773l-3.492 1.836 0.667-3.888-2.825-2.753 3.904-0.567 1.746-3.537 1.746 3.537 3.904 0.567-2.825 2.753 0.667 3.888-3.492-1.836z"
        />
      </svg>
      <svg
        fill="currentColor"
        height="15"
        stroke="currentColor"
        stroke-width="0"
        version="1.1"
        viewBox="0 0 16 16"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 6.204l-5.528-0.803-2.472-5.009-2.472 5.009-5.528 0.803 4 3.899-0.944 5.505 4.944-2.599 4.944 2.599-0.944-5.505 4-3.899zM8 11.773l-3.492 1.836 0.667-3.888-2.825-2.753 3.904-0.567 1.746-3.537 1.746 3.537 3.904 0.567-2.825 2.753 0.667 3.888-3.492-1.836z"
        />
      </svg>
    </div>
  </div>
   
</DocumentFragment>
`;
